<a href="https://colab.research.google.com/github/mdrizvi-106/llm-portfolio-fraud-detection/blob/main/llm_portfolio.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd

In [ ]:
import kagglehub, os
path = kagglehub.dataset_download("mlg-ulb/creditcardfraud")
df = pd.read_csv(os.path.join(path, "creditcard.csv"))

sample = df.sample(5000)

# Build a plain-English summary to send to the LLM
summary = f"""
You are a financial compliance analyst. Below is a sample of 50 credit card transactions.

Key stats:
- Total transactions in sample: {len(sample)}
- Fraudulent transactions: {sample['Class'].sum()}
- Average transaction amount: £{sample['Amount'].mean():.2f}
- Max transaction amount: £{sample['Amount'].max():.2f}
- Min transaction amount: £{sample['Amount'].min():.2f}

Top 5 highest-value transactions:
{sample.nlargest(5, 'Amount')[['Amount', 'Class']].to_string(index=False)}

Based on this data, flag any anomalies you would investigate and explain your reasoning.
"""

print(summary)

Using Colab cache for faster access to the 'creditcardfraud' dataset.

You are a financial compliance analyst. Below is a sample of 50 credit card transactions.

Key stats:
- Total transactions in sample: 5000
- Fraudulent transactions: 5
- Average transaction amount: £88.40
- Max transaction amount: £5714.58
- Min transaction amount: £0.00

Top 5 highest-value transactions:
 Amount  Class
5714.58      0
5130.30      0
4171.80      0
3759.39      0
2718.84      0

Based on this data, flag any anomalies you would investigate and explain your reasoning.



In [ ]:
from groq import Groq

client = Groq(api_key="GROQ_API_KEY")

response = client.chat.completions.create(
    model="llama-3.1-8b-instant",
    messages=[
        {"role": "system", "content": "You are a financial fraud analyst. Be concise and structured."},
        {"role": "user", "content": summary}
    ]
)

result = response.choices[0].message.content
print(result)

**Anomaly Identification and Investigation**

As a financial compliance analyst, I will use the following criteria to flag potential anomalies:

1. **High-value transactions**: Transactions exceeding the average transaction amount significantly.
2. **Unusual transaction patterns**: Transactions with no clear merchant or location information.
3. **High-frequency transaction patterns**: Transactions within a short timeframe with the same characteristics.

Based on the provided data, I would flag the following anomalies for investigation:

1. **All high-value transactions**:
	* 5 transactions with amounts exceeding £4,000 (e.g., £5130.30, £4171.80, £3759.39).
2. **Transaction with no clear merchant**: Although merchant information is not provided, I would investigate the transaction with the maximum amount (£5714.58) to determine the merchant associated with this transaction and verify if it's legitimate.
3. **Transaction with high-value and low-average-amounts**: The transactions with am

**Anomaly Identification and Investigation**

As a financial compliance analyst, I will use the following criteria to flag potential anomalies:

1. **High-value transactions**: Transactions exceeding the average transaction amount significantly.
2. **Unusual transaction patterns**: Transactions with no clear merchant or location information.
3. **High-frequency transaction patterns**: Transactions within a short timeframe with the same characteristics.

Based on the provided data, I would flag the following anomalies for investigation:

1. **All high-value transactions**:
	* 5 transactions with amounts exceeding £4,000 (e.g., £5130.30, £4171.80, £3759.39).
2. **Transaction with no clear merchant**: Although merchant information is not provided, I would investigate the transaction with the maximum amount (£5714.58) to determine the merchant associated with this transaction and verify if it's legitimate.
3. **Transaction with high-value and low-average-amounts**: The transactions with amounts £3759.39, £4171.80, and £5130.30 have significantly higher values than the average transaction amount (£88.40), which suggests a possible manipulation of transaction values.

**Recommendations for Further Investigation:**

1. Verify the legitimacy of the high-value transactions.
2. Review the merchant information associated with these transactions.
3. Analyze the transaction data to identify any patterns or irregularities.
4. Cross-reference the transaction data with external sources, such as merchant databases, to identify potential red flags.

**Risk Level:**
High risk. The presence of high-value transactions and lack of merchant information raises concerns about potential financial irregularities.